AttributeError: 'Attribute' object has not attribute 'name'

<!--
  Hi there! Thank you for discovering and submitting an issue.

  Before you submit this, make sure that the issue doesn't already exist
  or if it is not closed.

  Is your issue fixed on the preview release?:
    pip install pylint astroid --pre -U
-->

Seems to be an issued directly related to PR https://github.com/PyCQA/pylint/pull/3998 and issue https://github.com/PyCQA/pylint/issues/3742. When using a metaclass from a subpackage that can't be imported, and then referencing another variable from said subpackage within the same scope, an ``AttributeError`` is raised. Removing the ``assert`` line in the given file below prevents the error. Came across this while writing unit tests, hence the ``assert`` statement.

### Steps to reproduce

Given a file `lintme.py`:
```python
# pylint: disable=C0304,C0114,C0116,C0115,R0903
import pkg.sub

class Failed(metaclass=pkg.sub.Metaclass):
    pass
assert pkg.sub.value is None
```

### Current behavior

Result of `pylint lintme.py`:
```
(pylint-bug) [rydrew@redhat-worklp pylint-error]$  pylint lintme.py
************* Module lintme
lintme.py:2:0: E0401: Unable to import 'pkg.sub' (import-error)
Exception on node <FunctionDef.test l.4 at 0x7f867ba766a0> in file '/home/rydrew/Documents/pylint-error/lintme.py'
Traceback (most recent call last):
  File "/home/rydrew/.pyenv/versions/pylint-bug/bin/pylint", line 8, in <module>
    sys.exit(run_pylint())
  File "/home/rydrew/.pyenv/versions/pylint-bug/lib64/python3.9/site-packages/pylint/__init__.py", line 24, in run_pylint
    PylintRun(sys.argv[1:])
  File "/home/rydrew/.pyenv/versions/pylint-bug/lib64/python3.9/site-packages/pylint/lint/run.py", line 384, in __init__
    linter.check(args)
  File "/home/rydrew/.pyenv/versions/pylint-bug/lib64/python3.9/site-packages/pylint/lint/pylinter.py", line 973, in check
    self._check_files(
  File "/home/rydrew/.pyenv/versions/pylint-bug/lib64/python3.9/site-packages/pylint/lint/pylinter.py", line 1007, in _check_files
    self._check_file(get_ast, check_astroid_module, name, filepath, modname)
  File "/home/rydrew/.pyenv/versions/pylint-bug/lib64/python3.9/site-packages/pylint/lint/pylinter.py", line 1033, in _check_file
    check_astroid_module(ast_node)
  File "/home/rydrew/.pyenv/versions/pylint-bug/lib64/python3.9/site-packages/pylint/lint/pylinter.py", line 1170, in check_astroid_module
    retval = self._check_astroid_module(
  File "/home/rydrew/.pyenv/versions/pylint-bug/lib64/python3.9/site-packages/pylint/lint/pylinter.py", line 1215, in _check_astroid_module
    walker.walk(ast_node)
  File "/home/rydrew/.pyenv/versions/pylint-bug/lib64/python3.9/site-packages/pylint/utils/ast_walker.py", line 77, in walk
    self.walk(child)
  File "/home/rydrew/.pyenv/versions/pylint-bug/lib64/python3.9/site-packages/pylint/utils/ast_walker.py", line 79, in walk
    callback(astroid)
  File "/home/rydrew/.pyenv/versions/pylint-bug/lib64/python3.9/site-packages/pylint/checkers/variables.py", line 853, in leave_functiondef
    self._check_metaclasses(node)
  File "/home/rydrew/.pyenv/versions/pylint-bug/lib64/python3.9/site-packages/pylint/checkers/variables.py", line 2098, in _check_metaclasses
    consumed.extend(self._check_classdef_metaclasses(child_node, node))
  File "/home/rydrew/.pyenv/versions/pylint-bug/lib64/python3.9/site-packages/pylint/checkers/variables.py", line 2142, in _check_classdef_metaclasses
    name = klass._metaclass.expr.name
AttributeError: 'Attribute' object has no attribute 'name'
```

### Expected behavior

Only the import error is shown

```
(pylint-bug) [rydrew@redhat-worklp pylint-error]$  pylint lintme.py
************* Module lintme
lintme.py:2:0: E0401: Unable to import 'pkg.sub' (import-error)
```

### pylint --version output

Result of `pylint --version` output:

```
pylint 3.0.0-a4
astroid 2.6.2
Python 3.9.5 (default, May 14 2021, 00:00:00) 
[GCC 11.1.1 20210428 (Red Hat 11.1.1-1)]
```

<!--
Additional dependencies:
```
pandas==0.23.2
marshmallow==3.10.0
...
```
-->

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -2109,7 +2109,6 @@ class VariablesChecker(BaseChecker):
 
         consumed = []  # [(scope_locals, consumed_key)]
         metaclass = klass.metaclass()
-
         name = None
         if isinstance(klass._metaclass, astroid.Name):
             name = klass._metaclass.name
@@ -2130,18 +2129,17 @@ class VariablesChecker(BaseChecker):
                 if found:
                     consumed.append((scope_locals, name))
                     break
-        if found is None and not metaclass:
-            current_expr = klass._metaclass
-            while isinstance(current_expr, astroid.Attribute):
-                current_expr = current_expr.expr
-            name = current_expr.name
-            if not (
+        if (
+            found is None
+            and not metaclass
+            and not (
                 name in astroid.Module.scope_attrs
                 or utils.is_builtin(name)
                 or name in self.config.additional_builtins
                 or name in parent_node.locals
-            ):
-                self.add_message("undefined-variable", node=klass, args=(name,))
+            )
+        ):
+            self.add_message("undefined-variable", node=klass, args=(name,))
 
         return consumed
 

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -2139,8 +2139,10 @@ class VariablesChecker(BaseChecker):
                 isinstance(klass._metaclass, astroid.Attribute)
                 and klass._metaclass.expr
             ):
-                name = klass._metaclass.expr.name
-
+                if isinstance(klass._metaclass.expr, astroid.Name):
+                    name = klass._metaclass.expr.name
+                elif isinstance(klass._metaclass.expr, astroid.Attribute):
+                    name = klass._metaclass.expr.attrname
             if name is not None:
                 if not (
                     name in astroid.Module.scope_attrs

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -2130,27 +2130,18 @@ class VariablesChecker(BaseChecker):
                 if found:
                     consumed.append((scope_locals, name))
                     break
-
         if found is None and not metaclass:
-            name = None
-            if isinstance(klass._metaclass, astroid.Name):
-                name = klass._metaclass.name
-            elif (
-                isinstance(klass._metaclass, astroid.Attribute)
-                and klass._metaclass.expr
+            current_expr = klass._metaclass
+            while isinstance(current_expr, astroid.Attribute):
+                current_expr = current_expr.expr
+            name = current_expr.name
+            if not (
+                name in astroid.Module.scope_attrs
+                or utils.is_builtin(name)
+                or name in self.config.additional_builtins
+                or name in parent_node.locals
             ):
-                if isinstance(klass._metaclass.expr, astroid.Name):
-                    name = klass._metaclass.expr.name
-                elif isinstance(klass._metaclass.expr, astroid.Attribute):
-                    name = klass._metaclass.expr.attrname
-            if name is not None:
-                if not (
-                    name in astroid.Module.scope_attrs
-                    or utils.is_builtin(name)
-                    or name in self.config.additional_builtins
-                    or name in parent_node.locals
-                ):
-                    self.add_message("undefined-variable", node=klass, args=(name,))
+                self.add_message("undefined-variable", node=klass, args=(name,))
 
         return consumed
 

--- a/tests/functional/r/regression/regression_4680.py
+++ b/tests/functional/r/regression/regression_4680.py
@@ -1,17 +1,17 @@
-# pylint: disable=missing-docstring,too-few-public-methods,undefined-variable
+# pylint: disable=missing-docstring,too-few-public-methods
 
 import pkg.sub  # [import-error]
 
 
-class Failed(metaclass=pkg.sub.Metaclass):
+class Failed(metaclass=pkg.sub.Metaclass):  # [undefined-variable]
     pass
 
 
-class FailedTwo(metaclass=ab.ABCMeta):
+class FailedTwo(metaclass=ab.ABCMeta):  # [undefined-variable]
     pass
 
 
-class FailedThree(metaclass=pkg.sob.Metaclass):
+class FailedThree(metaclass=pkg.sob.Metaclass):  # [undefined-variable]
     pass
 
 

--- a/tests/functional/r/regression/regression_4680.py
+++ b/tests/functional/r/regression/regression_4680.py
@@ -1,0 +1,18 @@
+# pylint: disable=missing-docstring,too-few-public-methods,undefined-variable
+
+import pkg.sub  # [import-error]
+
+
+class Failed(metaclass=pkg.sub.Metaclass):
+    pass
+
+
+class FailedTwo(metaclass=ab.ABCMeta):
+    pass
+
+
+class FailedThree(metaclass=pkg.sob.Metaclass):
+    pass
+
+
+assert pkg.sub.value is None

--- a/tests/functional/r/regression/regression_4680.py
+++ b/tests/functional/r/regression/regression_4680.py
@@ -3,7 +3,7 @@
 import pkg.sub  # [import-error]
 
 
-class Failed(metaclass=pkg.sub.Metaclass):  # [undefined-variable]
+class Failed(metaclass=pkg.sub.Metaclass):
     pass
 
 
@@ -11,7 +11,7 @@ class FailedTwo(metaclass=ab.ABCMeta):  # [undefined-variable]
     pass
 
 
-class FailedThree(metaclass=pkg.sob.Metaclass):  # [undefined-variable]
+class FailedThree(metaclass=pkg.sob.Metaclass):
     pass
 
 

--- a/tests/functional/r/regression/regression_4680.txt
+++ b/tests/functional/r/regression/regression_4680.txt
@@ -1,4 +1,2 @@
 import-error:3:0::Unable to import 'pkg.sub':HIGH
-undefined-variable:6:0:Failed:Undefined variable 'sub':HIGH
 undefined-variable:10:0:FailedTwo:Undefined variable 'ab':HIGH
-undefined-variable:14:0:FailedThree:Undefined variable 'sob':HIGH

--- a/tests/functional/r/regression/regression_4680.txt
+++ b/tests/functional/r/regression/regression_4680.txt
@@ -1,0 +1,1 @@
+import-error:3:0::Unable to import 'pkg.sub':HIGH

--- a/tests/functional/r/regression/regression_4680.txt
+++ b/tests/functional/r/regression/regression_4680.txt
@@ -1,1 +1,4 @@
 import-error:3:0::Unable to import 'pkg.sub':HIGH
+undefined-variable:6:0:Failed:Undefined variable 'sub':HIGH
+undefined-variable:10:0:FailedTwo:Undefined variable 'ab':HIGH
+undefined-variable:14:0:FailedThree:Undefined variable 'sob':HIGH

